speechbrain/core.py

```diff
@@ -1050,60 +1068,89 @@ def fit_batch(self, batch):
         -------
         detached loss
         """
-        valid_loss = False
-
-        # Managing automatic mixed precision
-        if self.auto_mix_prec:
-            with torch.autocast(device_type=torch.device(self.device).type):
-                outputs = self.compute_forward(batch, Stage.TRAIN)
-
-            # Losses are excluded from mixed precision to avoid instabilities
-            loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
-
-            if self.check_gradients(loss):
-                valid_loss = True
-                self.valid_step += 1
-
-            should_step = self.valid_step % self.grad_accumulation_factor == 0
-            if valid_loss:
-                with self.no_sync(not should_step):
-                    self.scaler.scale(
-                        loss / self.grad_accumulation_factor
-                    ).backward()
-                if should_step:
-                    self.scaler.unscale_(self.optimizer)
-                    self.scaler.step(self.optimizer)
-                    self.scaler.update()
-                    self.zero_grad()
-                    self.optimizer_step += 1
-        else:
-            if self.bfloat16_mix_prec:
-                with torch.autocast(
-                    device_type=torch.device(self.device).type,
-                    dtype=torch.bfloat16,
-                ):
-                    outputs = self.compute_forward(batch, Stage.TRAIN)
-                    loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
-            else:
-                outputs = self.compute_forward(batch, Stage.TRAIN)
-                loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
-
-            if self.check_gradients(loss):
-                valid_loss = True
-                self.valid_step += 1
-
-            should_step = self.valid_step % self.grad_accumulation_factor == 0
-            if valid_loss:
-                with self.no_sync(not should_step):
-                    (loss / self.grad_accumulation_factor).backward()
-                if should_step:
-                    self.optimizer.step()
-                    self.zero_grad()
-                    self.optimizer_step += 1
+        amp = AutocastConfig.from_name(self.autocast)
+        should_step = (self.step % self.grad_accumulation_factor) == 0
+
+        # We wrap the forward pass in no_sync to avoid
+        # unnecessary synchronization of gradients.
+        # Empirically, we found that this improves performance by a lot (2x speedup).
+        with self.no_sync():
+            with torch.autocast(
+                dtype=amp.dtype, device_type=torch.device(self.device).type,
+            ):
+                outputs = self.compute_forward(batch, sb.Stage.TRAIN)
+
+        with self.no_sync(not should_step):
```

Comment:
Referring to [this](https://github.com/Adel-Moumen/speechbrain/blob/c1a23a1a672ac7c8f7bb270b99b7282daa9cfa72/speechbrain/core.py#L1600), the current version of `no_sync` leads to a silent bug when using DDP + gradient accumulation. We `break` from the loop as soon as we find a module without `require_backward_grad_sync`, i.e. not wrapped in DDP:

```python
if use:
    old_values_list = []
    for module in self.modules.values():
        if not hasattr(module, "require_backward_grad_sync"):
            # if not using DDP
            break
        old_values_list.append(module.require_backward_grad_sync)
        module.require_backward_grad_sync = False
    yield
    for module, old_value in zip(
        self.modules.values(), old_values_list
    ):
        if not hasattr(module, "require_backward_grad_sync"):
            break
        module.require_backward_grad_sync = old_value
else:
    yield
```

The assumption here is that either all modules are wrapped in DDP or none of them. But this is not the case in the majority of the recipes. We typically include in the YAML `modules` the feature extractor, the global normalization, the augmentation, frozen wav2vec2 etc., which are `torch.nn.Module` but without parameters or with all parameters frozen. These modules are not wrapped in DDP (see [here](https://github.com/Adel-Moumen/speechbrain/blob/c1a23a1a672ac7c8f7bb270b99b7282daa9cfa72/speechbrain/core.py#L1484)). Therefore `no_sync` unsets `require_backward_grad_sync` for each DDP module until a non-DDP module is found while iterating `self.modules` => when using DDP + gradient accumulation, some gradients are correctly synchronized every `grad_accumulation_factor` steps, other gradients are unnecessarily synchronized at each step. This can cause slowdowns and maybe even convergence issues. Something like this should fix the issue:

```python
if use:
    old_values_list = []
    for module in self.modules.values():
        if not hasattr(module, "require_backward_grad_sync"):
            # if not using DDP
            continue
        old_values_list.append(module.require_backward_grad_sync)
        module.require_backward_grad_sync = False
    yield
    i = 0
    for module in self.modules.values():
        if not hasattr(module, "require_backward_grad_sync"):
            continue
        module.require_backward_grad_sync = old_values_list[i]
        i += 1
else:
    yield
```

Reply:
I think you are right about convergence issues. See the following tests that I did with CommonVoice FR + w2v2 + 2x GPUs (rtx 2080 ti):
DDP with no_sync(not should_step): 
  6%|██████▊                                                                                                                  | 999/17829 [11:10<3:08:17,  1.49it/s, train_loss=2.17]
DDP with no_sync(not should_step) + fix 
  6%|██████▊                                                                                                                  | 999/17829 [11:13<3:09:14,  1.48it/s, train_loss=2.15]
DDP with no_sync()
  6%|██████▊                                                                                                                  | 999/17829 [10:05<2:50:01,  1.65it/s, train_loss=3.62]

I stopped all training at 999 steps. The first is when we are wrapping the fwd inside the no_sync(not should_step) without your fix, the second test is with wrapping + your fix, and the last one is what I initially proposed (i.e. fwd in self.no_sync()). 

We can see that in the first two cases, the train loss is much lower (2.17) than the last one (3.62). However, I'm not seeing any improvements in terms of speedup yet…

Reply:
With 1x rtx 2080 ti:
  6%|██████▋                                                                                                                 | 1999/35658 [20:14<5:40:57,  1.65it/s, train_loss=1.53]
  
  I stopped at 1999 to simulate the same number of batches.

Reply:
So it seems that having the `no_sync` is a bit faster but not that much. However, it yields to a very bad train loss.